### PR TITLE
CDR config API 2.20 changes

### PR DESCRIPTION
**Pull request by VindhyaRajagopal**
## The problem ##
[WA-3633](https://frollo.atlassian.net/browse/WA-3633)

## Checklist ##

Use the :white_check_mark: emoji for when it's been done; :no_entry: when it still needs work;
strikethrough when an item of the checklist does not apply to the changes.

* Is the destination branch correct & intended? :white_check_mark:
* Is the code synced to latest (if applicable)? :white_check_mark:
* Has the SDK example app build passed locally? :white_check_mark:
* Is the correct JIRA ticket referenced? :white_check_mark:
* Is the correct API docs/Confluence Page referenced? :white_check_mark:
* Have all the AC's in the ticket passed? :white_check_mark:
* Have you followed proper naming conventions in the code? :white_check_mark:
* Has the code been properly covered with unit & android tests?NA
* Have the full test suite (unit & android tests) passed locally?NA
* Has the changelog.md, packages.md, overview.md, extras.md been updated? :white_check_mark:
* Has the dokka task passed?NA
* Have you committed database schema changes (if applicable)? :white_check_mark:

### CODE REVIEW

**Review by hailefrollo — APPROVED** [review]

File: src/main/kotlin/us/frollo/frollosdk/aggregation/Aggregation.kt [hunk 1]
```diff
@@ -2322,6 +2322,7 @@ class Aggregation(network: NetworkService, internal val db: SDKDatabase, localBr
         after: Long? = null,
         before: Long? = null,
         size: Long? = null,
+        cdrConfigExternalId: String? = null,
```
Comment (deepakmn86) [inline src/main/kotlin/us/frollo/frollosdk/aggregation/Aggregation.kt:2325]: Actually i forgot why we added this. I didn't find this query parameter in the API documentation

File: changelog.md [hunk 2]
```diff
@@ -1,5 +1,11 @@
 ## Changelog
 
+### 3.20.0
+
+#### Changes
+- Add/Remove fields in CDR Config API Response
+- Add field in ConsentResponse & request
```
Comment (deepakmn86) [inline changelog.md:7]: this should be part of 3.19 changes only below.

**Review by deepakmn86 — CHANGES_REQUESTED** [review]

File: src/main/kotlin/us/frollo/frollosdk/database/SDKDatabase.kt [hunk 3]
```diff
@@ -766,18 +766,34 @@ abstract class SDKDatabase : RoomDatabase() {
...
-                // 1) Alter table - cdr_configuration - Add column initial_sync_window_weeks
+                // 1) Alter table - cdr_configuration - Add column initial_sync_window_weeks, software_id, software_name, summary, description
                 // 2) Alter table - message - Add columns created_date, delivered_date, interacted_date
+                // 3) Alter table - cdr_configuration - Rename related_parties to parties
+                // 4) Drop & re-create - cdr_configuration - Delete adr_id, adr_name columns and Add software_id, software_name, summary, description
+                // 5) Alter table - consent - Add column cdr_config_external_id
 
                 // START - Add column initial_sync_window_weeks
                 database.execSQL("ALTER TABLE `cdr_configuration` ADD COLUMN `initial_sync_window_weeks` INTEGER")
                 // END - Add column initial_sync_window_weeks
```
Comment (deepakmn86) [inline src/main/kotlin/us/frollo/frollosdk/database/SDKDatabase.kt:777]: delete above 3

File: src/main/kotlin/us/frollo/frollosdk/database/SDKDatabase.kt [hunk 4]
```diff
@@ -766,18 +766,34 @@ abstract class SDKDatabase : RoomDatabase() {
...
 
+                // START - Drop column adr_id, adr_name and add software_id, software_name, summary, description
```
Comment (deepakmn86) [inline src/main/kotlin/us/frollo/frollosdk/database/SDKDatabase.kt:779]:
```
// START - Drop columns adr_id, adr_name and Add columns initial_sync_window_weeks,
// additional_permissions, software_id, software_name, summary, image_url, description
```

File: src/main/kotlin/us/frollo/frollosdk/database/SDKDatabase.kt [hunk 5]
```diff
@@ -766,18 +766,34 @@ abstract class SDKDatabase : RoomDatabase() {
...
+                database.execSQL("INSERT INTO cdr_configuration(config_id, support_email, sharing_durations, permissions, additional_permissions, external_id, display_name, cdr_policy_url, model, related_parties, sharing_use_duration, initial_sync_window_weeks, software_id, software_name, image_url, summary, description) SELECT config_id, support_email, sharing_durations, permissions, additional_permissions, external_id, display_name, cdr_policy_url, model, related_parties, sharing_use_duration, initial_sync_window_weeks, software_id, software_name, image_url, summary, description FROM cdr_configuration")
+                database.execSQL("DROP TABLE original_cdr_configuration")
+                database.execSQL("CREATE INDEX IF NOT EXISTS `index_cdr_configuration_config_id` ON `cdr_configuration` (`config_id`)")
+                // End - Drop column adr_id, adr_name and add software_id, software_name, summary, description
```
Comment (deepakmn86) [inline src/main/kotlin/us/frollo/frollosdk/database/SDKDatabase.kt:786]:
```
// END - Drop columns adr_id, adr_name and Add columns initial_sync_window_weeks,
// additional_permissions, software_id, software_name, summary, image_url, description
```

File: src/main/kotlin/us/frollo/frollosdk/database/SDKDatabase.kt [hunk 6]
```diff
@@ -766,18 +766,34 @@ abstract class SDKDatabase : RoomDatabase() {
...
+                database.execSQL("CREATE TABLE IF NOT EXISTS `cdr_configuration` (`config_id` INTEGER NOT NULL, `support_email` TEXT NOT NULL, `sharing_durations` TEXT NOT NULL, `permissions` TEXT, `additional_permissions` TEXT, `external_id` TEXT NOT NULL, `display_name` TEXT NOT NULL, `cdr_policy_url` TEXT NOT NULL, `model` TEXT NOT NULL, `related_parties` TEXT NOT NULL, `sharing_use_duration` INTEGER NOT NULL, `initial_sync_window_weeks` INTEGER, `software_id` TEXT, `software_name` TEXT, `image_url` TEXT, `summary` TEXT, `description` TEXT, PRIMARY KEY(`config_id`))")
+                database.execSQL("INSERT INTO cdr_configuration(config_id, support_email, sharing_durations, permissions, additional_permissions, external_id, display_name, cdr_policy_url, model, related_parties, sharing_use_duration, initial_sync_window_weeks, software_id, software_name, image_url, summary, description) SELECT config_id, support_email, sharing_durations, permissions, additional_permissions, external_id, display_name, cdr_policy_url, model, related_parties, sharing_use_duration, initial_sync_window_weeks, software_id, software_name, image_url, summary, description FROM cdr_configuration")
```
Comment (deepakmn86) [inline src/main/kotlin/us/frollo/frollosdk/database/SDKDatabase.kt:783]: as original table doesn't have these new columns we cannot have these columns in the insert command. and also we have to insert from original table to the newly created table.

```
database.execSQL("INSERT INTO cdr_configuration(config_id, support_email, sharing_durations, permissions, external_id, display_name, cdr_policy_url, model, related_parties, sharing_use_duration) SELECT config_id, support_email, sharing_durations, permissions, external_id, display_name, cdr_policy_url, model, related_parties, sharing_use_duration FROM original_cdr_configuration")
```

**Review by deepakmn86 — CHANGES_REQUESTED** [review]

**Review by deepakmn86 — APPROVED** [review]

**Pull request merged**